### PR TITLE
Adds Autoimplanter for nuke ops

### PR DESCRIPTION
**Pull request by PKPenguin321**
The Autoimplanter accepts any cyber implants (such as xray/thermal/hud and reviver/nutrient/CNS/nodrop implants). When used in your hand, it will instantly implant you with whatever cyberimplant was inserted into it. Implants that were inserted accidentally can be removed with a screwdriver.

The box of implants that nuke ops can purchase will now always come with an autoimplanter.

The sprite for reference, by I believe @newfren 
rocsczo

:cl: PKPenguin321
rscadd: The Autoimplanter, a device that can insert cyberimplants into humans instantly and without the need of surgery, has been added to the game. It is currently only obtainable by nuke ops, by way of being included in the Box of Implants.
/:cl:

### CODE REVIEW

File: code/modules/surgery/organs/autoimplanter.dm [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	storedorgan.Insert(user)//insert stored organ into the user
+	user << "<span class='notice'>You feel a slight tickle as [src] quickly implants you with the [storedorgan].</span>"
+	storedorgan = null
+
+/obj/item/device/autoimplanter/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/organ/internal/cyberimp))
+		if(storedorgan)
+			user << "<span class='notice'>[src] already has an implant stored.</span>"
+			return
+		user.unEquip(I)
```
Comment (neersighted) [inline code/modules/surgery/organs/autoimplanter.dm]:
```
if(!user.drop_item())
    return
```

File: code/modules/surgery/organs/autoimplanter.dm [hunk 2]
```diff
@@ -0,0 +1,33 @@
+/obj/item/device/autoimplanter
+	name = "autoimplanter"
+	desc = "A device that automatically injects a cyber-implant into the user without the hassle of extensive surgery. It has a slot to insert implants and a screwdriver slot for removing accidentally added implants."
+	icon_state = "gangtool-white"
```
Comment (neersighted) [inline code/modules/surgery/organs/autoimplanter.dm]: Unique sprites would be good.

Comment (PKPenguin321) [inline code/modules/surgery/organs/autoimplanter.dm]: i'll make a post in the sprites request thread, but no promises

File: code/modules/surgery/organs/autoimplanter.dm [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+	item_state = "walkietalkie"
+	w_class = 2
+	var/obj/item/organ/internal/cyberimp/storedorgan
+
+/obj/item/device/autoimplanter/attack_self(mob/user, obj/item/I)//when the object it used...
+	if(!storedorgan)
+		user << "<span class='notice'>[src] currently has no implant stored.</span>"
+		return
+	storedorgan.Insert(user)//insert stored organ into the user
+	user << "<span class='notice'>You feel a slight tickle as [src] quickly implants you with the [storedorgan].</span>"
```
Comment (neersighted) [inline code/modules/surgery/organs/autoimplanter.dm]: `visible_message` maybe? Others probably see this.

File: code/modules/surgery/organs/autoimplanter.dm [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+/obj/item/device/autoimplanter/attackby(obj/item/I, mob/user, params)
+	if(istype(I, /obj/item/organ/internal/cyberimp))
+		if(storedorgan)
+			user << "<span class='notice'>[src] already has an implant stored.</span>"
+			return
+		user.unEquip(I)
+		I.loc = src
+		storedorgan = I
+		user << "<span class='notice'>You insert the [I] into [src].</span>"
+	if(istype(I, /obj/item/weapon/screwdriver))
```
Comment (neersighted) [inline code/modules/surgery/organs/autoimplanter.dm]: `else if`

File: code/modules/surgery/organs/autoimplanter.dm (hunk 2 above)
Comment (neersighted) [inline code/modules/surgery/organs/autoimplanter.dm]: You can always beg @WJohn

**Comment by xxalpha** [comment]
Personally I feel it would be more intuitive if attack_self() was used for removing an implant from the implanter and implating would be done on attacking yourself or another person.

File: code/modules/surgery/organs/autoimplanter.dm [hunk 5]
```diff
@@ -0,0 +1,33 @@
+/obj/item/device/autoimplanter
+	name = "autoimplanter"
+	desc = "A device that automatically injects a cyber-implant into the user without the hassle of extensive surgery. It has a slot to insert implants and a screwdriver slot for removing accidentally added implants."
+	icon_state = "gangtool-white"
+	item_state = "walkietalkie"
+	w_class = 2
+	var/obj/item/organ/internal/cyberimp/storedorgan
+
+/obj/item/device/autoimplanter/attack_self(mob/user, obj/item/I)//when the object it used...
```
Comment (xxalpha) [inline code/modules/surgery/organs/autoimplanter.dm]: ``attack_self()`` shouldn't have ``obj/item/I`` argument.

**Comment by GunHog** [comment]
Oh man, this is good! If Ops did not explode on death, I would so steal this as RD for mass self implanting! It is better than having an AutoDoc machine!

File: code/modules/surgery/organs/autoimplanter.dm (hunk 5 above)
Comment (PKPenguin321) [inline code/modules/surgery/organs/autoimplanter.dm]: shit you're right, i believe we figured this out earlier but i forgot to tweak it in the finalized code

**Comment by PKPenguin321** [comment]
@GunHog don't you dare add this to R&D

File: code/modules/surgery/organs/autoimplanter.dm (hunk 4 above)
Comment (xxalpha) [inline code/modules/surgery/organs/autoimplanter.dm]: I suggest returning at the end of the first if block instead.

File: code/modules/surgery/organs/autoimplanter.dm (hunk 1 above)
Comment (PKPenguin321) [inline code/modules/surgery/organs/autoimplanter.dm]: do i replace this line with that or add that under this line? also, what exactly does it do?

Comment (neersighted) [inline code/modules/surgery/organs/autoimplanter.dm]: It checks that the user can drop the item (and drops it), or bails if they can't. Replace `unEquip` with it.

**Comment by GunHog** [comment]
You are such a fun hater, geez.

**Comment by PKPenguin321** [comment]
okay fine *maybe* you can add it to R&D, but it should require a bunch of illegal and bio tech. i don't *really* care that much but whatever

File: code/modules/surgery/organs/autoimplanter.dm [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+	item_state = "walkietalkie"
+	w_class = 2
+	var/obj/item/organ/internal/cyberimp/storedorgan
+
+/obj/item/device/autoimplanter/attack_self(mob/user)//when the object it used...
+	if(!storedorgan)
+		user << "<span class='notice'>[src] currently has no implant stored.</span>"
+		return
+	storedorgan.Insert(user)//insert stored organ into the user
+	user.visible_message("<span class='notice'>[user] presses a button on [src], and you hear a short mechanical noise.</span>")
```
Comment (neersighted) [inline code/modules/surgery/organs/autoimplanter.dm]: Oops, I meant to specify a message for both... For instance:
`user.visible_message("<span class='notice'>[user] presses a button on [src], and you hear a short mechanical noise.</span>", "<span class='notice'>You feel a sharp sting as [src] plunges into your body.</span>")`

Comment (PKPenguin321) [inline code/modules/surgery/organs/autoimplanter.dm]: i'm not 100% sure what you mean. you're saying i can put the message for everybody and the message for the user in the same line?

Comment (neersighted) [inline code/modules/surgery/organs/autoimplanter.dm]: Yes, `visible_message(everybody_else, you)`

**Comment by Lumipharon** [comment]
It's good for ops, but please don't just add a self surgery work around for regular play.

**Comment by PKPenguin321** [comment]
like i said, if that gets added itll be on gunhog, not me

**Comment by GunHog** [comment]
It is probably never going to happen (on this codebase). No matter what form or such, self-surgery will probably be rejected.

File: code/modules/surgery/organs/autoimplanter.dm [hunk 7]
```diff
@@ -0,0 +1,35 @@
+/obj/item/device/autoimplanter
+	name = "autoimplanter"
+	desc = "A device that automatically injects a cyber-implant into the user without the hassle of extensive surgery. It has a slot to insert implants and a screwdriver slot for removing accidentally added implants."
+	icon_state = "gangtool-white"
+	item_state = "walkietalkie"
```
Comment (Fox-McCloud) [inline code/modules/surgery/organs/autoimplanter.dm]: this doesn't even have an in-hand anywhere; pretty sure this item_state doesn't exist.

Comment (PKPenguin321) [inline code/modules/surgery/organs/autoimplanter.dm]: Correct and intentional until I get actual sprites. Not sure that I even need I hands for this, anyways.

Comment (Fox-McCloud) [inline code/modules/surgery/organs/autoimplanter.dm]: it's just a personal annoyance is all; this item_state is used a lot, but doesn't exist....blehhh

**Comment by newfren** [comment]
I might try to make sprites for this.

**Comment by Robertha89** [comment]
image
here you go, one giant implanter.

**Comment by Fox-McCloud** [comment]
that's....a bucket with some other sprite glued on.

**Comment by LanCartwright** [comment]
Here how about this:
A medibot that implants you.
Takes a long time. When emagged it knocks you over and mutilates your genitals.

**Comment by AlexanderUlanH** [comment]
Self-surgery would be fine as long as it was balanced by consequences.  "You can't implant yourself because you'd bleed out and die before you successfully performed all the steps" is a much better way of balancing than "You can't implant yourself because there's a check for target == usr"

But this is definitely needed regardless because ops simply don't have time for surgery.

**Comment by GunHog** [comment]
I would be happy to code some sort of AutoDoc for the Medbay Surgery room (But perhaps not provide one for Robotics, make them build it first?), it makes more sense than self surgery in my opinion.

That is unrelated to this PR, though. Even though Nuke Ops SHOULD be able to trust each other with surgeries, it does not seem to happen much.

**Comment by PKPenguin321** [comment]
the main reason for this is because nuke ops are slow enough getting out of their base as it is, and should not need to do 6 different organ manipulation surgeries and have to wait half an hour longer just because they bought an implant bundle.

**Comment by xxalpha** [comment]
This really should be a Nuke Ops tool exclusively.

**Comment by PKPenguin321** [comment]
it is

**Comment by xxalpha** [comment]
I am just saying...sooner or later someone is going to want to add it somewhere else.

**Comment by Dorsisdwarf** [comment]
Add a comment to it saying that "If you add this to R&D future coders I will be forced to slay you"

**Comment by PKPenguin321** [comment]
conflict is a sprite conflict in devices.dmi, will fix

**Comment by PKPenguin321** [comment]
yyyyeaahhhh i did it without blowing it all to hell! i am the best
@Razharas thank you for guiding me through resolving conflicts in git in the absence of oranges. now we can merge the PR, right? :)

**Comment by PKPenguin321** [comment]
@Razharas @RemieRichards @duncathan anybody with merge access do the mergey thing PLEASE

**Comment by xxalpha** [comment]
Beg in coderbus like a good goy.

**Comment by PKPenguin321** [comment]
>mfw my feature gets merged

ye

**Pull request merged**